# DNA Methylation in Ecology and Evolution

### Physalia-Courses 

# Short-reads workshop Day 2 // 11th February, 2025

This is the Readme file for the session on Tuesday, where we will cover the second half of the short-read processing. All the data you'll need for this will be stored in: 

`~/Share/2_Tuesday/`

## Overview

Using the same data as yesterday, we will; examine the bisulfite conversion rate, extract readable DNA methylation C/T count data and account for genetic variation caused by SNPs.

## 1. Examining the Bisulfite / Enzymatic conversion rate

For WGBS/RRBS a 1% spike of lambda phage genome is included in the DNA sample before bisulfite conversion. Lambda is unmethylated so acts as a negative control to understand how efficent the bisulfite converstion is. For EM-Seq both a lambda (negative control) and pUC19 (plasmid positive control - fully methylated) are used. It is also possible to check the conversation rate of samples by examining the chloroplast/mitochondria sequences as these are thought to be unmethylated. However, this has recently been questioned and so it is always better to use a controlled spike.

Today for ease, we will check the conversion rate of the Arabidopsis data using the chloroplast genome.

### Core Tasks:
* Copy the Arabidopsis chloroplast genome into your area
* Using what you learnt yesterday, prepare the genome using bismark
* Align your sequence files to the chloroplast genome
* Examine the output report - what is your conversion rate?

### Advanced Tasks:
* Write a loop to align multiple sequencing files
* Create a short Snakemake script to align multiple sequencing files
* Attempt this on your own HPC using your own data (or practice data from NCBI)

Whilst primarily checking the conversion rate ensures your data are of good quality. It is also important for later downstream processing of the data. We use this number as an error level when carrying out binomial tests to call positions as methylated/unmethylated.

*Other than recording the conversion rate, we do not use any of these files for further steps.*


## 2. Extracting Useable Methylation Counts

After creating `.bam` files by aligning the sequencing data to the species reference genome, we must now convert these files into useable dataframes of methylation information. This is simply achieved with a few more `Bismark` commands.

### Core Tasks:
* Examine the `Bismark` documentation: [https://felixkrueger.github.io/Bismark/](https://felixkrueger.github.io/Bismark/)
* Carry out deduplication of your `.bam` files
* Carry out `methylation_extraction` (see code below)

You will see there are lots of options for creating lots of different files depending on what you want to do with your data. For simple differential DNA methylation analysis, all we need is the coverage file. Generated by running the below step:

```
bismark_methylation_extractor --multicore 3 --comprehensive --bedGraph --report --cytosine_report \
--genome_folder TAIR10_genome/ <your_deduplicated>.bam
```

* Examine your M-bias plots by downloading the `.png` files to your computer
* Do you think you should run `methylation_extraction` again using the `--ignore` options?

Once you are happy with your final extracted files these can be used for differential DNA methylation analyses in R (Day 4). 


**Optional:**\
We can run one more command which merges the calls across strands. As CpGs consist of a C on both the positive and negative DNA strand, we can merge this information to create a single CpG position for later anaysis. This is a personal decision based on your research question, you may be interested in knowing what's happening on both strands, or not...

* Carry out the strand merge

```
coverage2cytosine -o <file_name> --merge_CpGs --genome_folder TAIR10_genome/ <your_deduplicated>.bam
```

These merged coverage files can also go straight into the differential DNA methylation pipeline.

### Advanced Tasks:
* Write a loop to carry out each step for all files
* Add the above steps to your Snakemake script
* Attempt this on your own HPC using your own data (or practice data from NCBI)

## 3. Accounting for SNPs

to-do ...
